Sources/XiEditor/EditView.swift

```diff
@@ -434,16 +425,14 @@ final class EditView: NSView, NSTextInputClient, TextPlaneDelegate {
         }
         let linespace = dataSource.textMetrics.linespace
         let topPad = linespace - dataSource.textMetrics.ascent
-        let xOff = dataSource.gutterWidth + x0 - scrollOrigin.x
-        let yOff = topPad - scrollOrigin.y
-        let firstVisible = max(0, Int((floor(dirtyRect.origin.y - topPad + scrollOrigin.y) / linespace)))
-        let lastVisible = max(0, Int(ceil((dirtyRect.origin.y + dirtyRect.size.height - topPad + scrollOrigin.y) / linespace)))
+        let xOff = dataSource.gutterWidth + x0 - dataSource.scrollOrigin.x
+        let yOff = topPad - dataSource.scrollOrigin.y
 
         // Note: this locks the line cache for the duration of the render
         let lineCache = dataSource.lines.locked()
         let totalLines = lineCache.height
-        let first = min(totalLines, firstVisible)
-        let last = min(totalLines, lastVisible)
+        let first = min(yOffsetToLine(dirtyRect.origin.y + dataSource.scrollOrigin.y), totalLines)
```

Comment:
Don't we need to use `topPad` value for `first` and `last` calculation?

Reply:
now I'm wondering if topPad should be part of the actual `yOffsetToLine` fn. I think yes.

Reply:
Okay, so after digging a bit deeper: `topPad` was handled in `yOffsetToLine`, sort of, by the use of `textMetrics.descent`. Assuming a font has no leading, descent == linespace - ascent. This was definitely confusing, though; I've added a new commit that just adds a `topPadding` calculated property to `TextDrawingMetrics`, so that this will be the same everywhere.